server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,330 @@
+                     *     ENTRY_MODIFY:settings.json
+                     * Linux in Docker symlink delete and set (rm -rf operator && ln -s <path to>/file_settings/ operator):
+                     *     ENTRY_CREATE:operator
+                     * Windows
+                     *     ENTRY_CREATE:operator
+                     *     ENTRY_MODIFY:operator
+                     * After we get an indication that something has changed, we check the timestamp, file id,
+                     * real path of our desired file.
+                     */
+                    if (Files.exists(settingsDir)) {
+                        try {
+                            Path path = operatorSettingsFile();
+
+                            if (logger.isDebugEnabled()) {
+                                key.pollEvents().stream().forEach(e -> logger.debug("{}:{}", e.kind().toString(), e.context().toString()));
+                            }
+
+                            key.pollEvents();
+                            key.reset();
+
+                            // We re-register the settings directory watch key, because we don't know
+                            // if the file name maps to the same native file system file id. Symlinks
+                            // are one potential cause of inconsistency here, since their handling by
+                            // the WatchService is platform dependent.
+                            settingsDirWatchKey = enableSettingsWatcher(settingsDirWatchKey, settingsDir);
+
+                            if (watchedFileChanged(path)) {
+                                processFileSettings(path, false);
+                            }
+                        } catch (Exception e) {
+                            logger.warn("error processing operator settings file", e);
+                        }
+                    } else {
+                        key.pollEvents();
+                        key.reset();
+                    }
+                }
+            } catch (ClosedWatchServiceException | InterruptedException expected) {
+                logger.info("shutting down watcher thread");
+            } catch (Exception e) {
+                logger.error("shutting down watcher thread with exception", e);
+            } finally {
+                watcherThreadLatch.countDown();
+            }
+        }, "elasticsearch[file-settings-watcher]").start();
+    }
+
+    synchronized void stopWatcher() {
+        logger.debug("stopping watcher ...");
+        if (watching()) {
+            try {
+                if (settingsDirWatchKey != null) {
+                    settingsDirWatchKey.cancel();
+                    settingsDirWatchKey = null;
+                }
+                fileUpdateState = null;
+                watchService.close();
+                if (watcherThreadLatch != null) {
+                    watcherThreadLatch.await();
+                }
+            } catch (IOException | InterruptedException e) {
```

Comment:
Stop happens on another ES thread, not one controlled here. So the interrupt should be restored for InterruptedException. (perhaps capture InterruptedException in another catch block). Note I think it is ok to not restore above since the thread is owned by this service and the thread will end after the interrupted exception is logged there.